### ISSUE TITLE
Rule: Keyword `new` should not be used

### ISSUE THREAD

**Issue by jissereitsma** [extdn/extdn-phpcs#17]
The `new` keyword should not be used to instantiate new classes. Constructor DI needs to be used instead.

**Comment by jissereitsma** [comment]
This was coined by @renttek in https://github.com/extdn/extdn-phpcs/issues/8 I personally do not agree with this, as this would limit non-Magento code too much. However, it is open for debate.

**Comment by avstudnitz** [comment]
I agree that there are some use cases when `new` is a good approach. It shouldn't be used for Magento classes though.

**Comment by fooman** [comment]
It's used heavily for exceptions:

              throw new \Magento\Framework\Exception\LocalizedException(
                    __('My Message here')
                );

How about for a data transport object?

        $transport = new \Magento\Framework\DataObject(
            ['test' => 'data']
        );
        $this->eventManager->dispatch(
            'custom_event',
            [
                'transport' => $transport
            ]
        );

**Comment by jissereitsma** [comment]
Completely agree with @fooman. Theoretically, exceptions and `DataObject` could be generated through factories as well, that are again injected in the constructor. However, this takes the DI pattern so literally, that it starts to be smelly.

**Comment by renttek** [comment]
I think the use of `new` is not bad per se. There are many cases like throwing Exceptions, object creation in Factories or for creating DTOs (like `\Magento\Framework\DataObject`) where `new` is absolutely ok.
I many other cases the usage of new creates side effects, which are also can make code more complex to test.
The question is, can/should this be checked?

**Comment by navarr** [comment]
Just wanted to add the bit that: Magento architects have said Exceptions should be created with the `new` keyword.  I got in trouble for creating an ExceptionFactory.

**Comment by schmengler** [comment]
We agreed that the spirit of the rule is good, namely to encourage dependency injection, but it may be too strict. We want to try it out for a while, with a low warning level and of course document when and why this makes sense.

**Comment by lenaorobei** [comment]
Might be helpful https://github.com/magento/marketplace-eqp/blob/master/MEQP2/Sniffs/Classes/ObjectInstantiationSniff.php

**Comment by schmengler** [comment]
Thank you @lenaorobei, I included and documented the rule in #45
